### PR TITLE
Issue/112

### LINKED ISSUE #112: Tidy Code

### PR DESCRIPTION
**Pull request by ravenmaster001**
## Proposed changes

This resolves #112. The changes contained are:

- [x] Code tidying, various bug fixes, and robustness fixes.
- [x] Flake8 Support.
- [x] Generic Processors.
- [x] Documentation for Generic Processors.

Tick each when checked/approved.

### CODE REVIEW

File: docs/source/contributing.rst [hunk 1]
```diff
@@ -102,11 +102,11 @@ The items of interest are:
...
+- ``Processors/`` - There is a different processor for each tag. A processor uses it's corresponding description loaded from ``processor-info.json`` to find matches in the text, and uses the given arguments in the matched tag to populate and output it's html template.
 
 - ``html-templates/`` - The html templates (using the Jinja2 template engine) with variable arguments to be populated by processors.
 
-- ``processor-info.json`` - Every processor is listed in this file, and will at least contain a regex pattern to match it's corresponding tag. Most will also define required and optional parameters, these correspond to arguments in the tag's html template.
+- ``processor-info.json`` - Every processor is listed in this file, and will at least contain a class determining where it is custom or generic, where custom processors will have a pattern to match it's corresponding tag. Most will also define required and optional parameters, these correspond to arguments in the tag's html template.
```
Comment (hayleyavw) [inline docs/source/contributing.rst]: > determining where it is custom or generic

or

> determining whether it is custom or generic

?

File: docs/source/contributing.rst [hunk 2]
```diff
@@ -117,7 +117,161 @@ It is important to note that Kordac is not just a Markdown Extension, it is a wr
...
+
+.. literalinclude:: ../../kordac/tests/assets/button-link/doc_example_basic_usage.md
+    :language: none
+
+To generate the output:
+
+.. literalinclude:: ../../kordac/tests/assets/button-link/doc_example_basic_usage_expected.html
+    :language: html
+
+A generic container processor, a pair of matching tags where one opens the container and one closes the container. The start tag gives the arguments as parameters to an html template. An the end tag is used to capture the content between the tags to be used as an additional parameter to the html template.  An example of a processor that uses the generic container processor is the :ref:`boxed-text <boxed-text>` processor which is described in the json as:
```
Comment (hayleyavw) [inline docs/source/contributing.rst]: > An the end tag

?

> The end tag

or

> An end tag

File: kordac/Kordac.py [hunk 3]
```diff
@@ -75,61 +76,64 @@ def convert(self, text):
...
+                mapping HTML strings as values.
                 eg: {'image': '<img src={{ source }}>'}
-        """
+        '''
         self.html_templates.update(html_templates)
         self.create_converter()
 
     def clear_templates(self):
-        """Set the template dictionary to it's original values."""
+        '''Set the template dictionary to it's original values.'''
```
Comment (hayleyavw) [inline kordac/Kordac.py]: trailing ''' should be on new line

File: kordac/Kordac.py [hunk 4]
```diff
@@ -75,61 +76,64 @@ def convert(self, text):
...
-        """Create a KordacResult object.
+        '''Create a KordacResult object.
 
         Args:
             html_string: A string of HTML text.
             title: The first heading encountered when converting.
-            required_files: Dictionary of required file types to sets of paths.
-        """
+            required_files: Dictionary of required file types to sets
+            of paths.
```
Comment (hayleyavw) [inline kordac/Kordac.py]: missing `heading_tree` and `required_glossary_terms`

File: kordac/KordacExtension.py [hunk 5]
```diff
@@ -107,17 +97,31 @@ def extendMarkdown(self, md, md_globals):
...
             self.required_files[key].clear()
 
     def loadJinjaTemplates(self, custom_templates):
+        '''Loads default templates from the templates directory, if
+        a custom template is given that will override the default
+        template.
+
+        Args:
+            custom_templates: a dictionary of names to custom templates
+                which are used to override default templates.
```
Comment (hayleyavw) [inline kordac/KordacExtension.py]: Returns?

File: kordac/processors/BeautifyPostprocessor.py [hunk 6]
```diff
@@ -4,7 +4,11 @@
 from bs4 import BeautifulSoup
 import re
 
+
 class BeautifyPostprocessor(Postprocessor):
+    ''' Converts the output document into a more asthetically
+    pleasing version with more consistent whitespace.
```
Comment (hayleyavw) [inline kordac/processors/BeautifyPostprocessor.py]: :ok_hand:

File: kordac/processors/ConditionalProcessor.py [hunk 7]
```diff
@@ -1,51 +1,121 @@
...
 
         Returns:
-            The next tag (regex match) the current block (string) and the content of the blocks (list of strings).
+            The next tag (regex match) the current block (string) and
+            the content of the blocks (list of strings).
 
         Raises:
-            TagNotMatchedError: When a sibling conditional is not closed.
+            TagNotMatchedError: When a sibling conditional is not
+            closed.
```
Comment (hayleyavw) [inline kordac/processors/ConditionalProcessor.py]: indent 118 (to match the way you've done the other docstrings so far)

File: kordac/processors/GenericContainerBlockProcessor.py [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+        self.template_parameters = ext.processor_info[self.processor].get('template_parameters', None)
+        self.process_parameters = lambda processor, parameters, argument_values: \
+            process_parameters(ext, processor, parameters, argument_values)
+
+    def test(self, parent, block):
+        ''' Tests a block to see if the run method should be applied.
+
+        Args:
+            parent: The parent node of the element tree that children
+            will reside in.
```
Comment (hayleyavw) [inline kordac/processors/GenericContainerBlockProcessor.py]: indent 29

File: kordac/processors/GenericContainerBlockProcessor.py [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+        return self.p_start.search(block) is not None or self.p_end.search(block) is not None
+
+    def run(self, parent, blocks):
+        ''' Generic run method for container tags.
+
+        Args:
+            parent: The parent node of the element tree that children
+            will reside in.
+            blocks: A list of strings of the document, where the
+            first block tests true.
```
Comment (hayleyavw) [inline kordac/processors/GenericContainerBlockProcessor.py]: indent 42, 44

File: kordac/processors/GenericTagBlockProcessor.py [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+        self.template_parameters = ext.processor_info[self.processor].get('template_parameters', None)
+        self.process_parameters = lambda processor, parameters, argument_values: \
+            process_parameters(ext, processor, parameters, argument_values)
+
+    def test(self, parent, block):
+        ''' Tests a block to see if the run method should be applied.
+
+        Args:
+            parent: The parent node of the element tree that children
+            will reside in.
```
Comment (hayleyavw) [inline kordac/processors/GenericTagBlockProcessor.py]: indent 30

File: kordac/processors/GenericTagBlockProcessor.py [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+        return self.pattern.search(block) is not None
+
+    def run(self, parent, blocks):
+        ''' Generic run method for single match tags.
+
+        Args:
+            parent: The parent node of the element tree that children
+            will reside in.
+            blocks: A list of strings of the document, where the
+            first block tests true.
```
Comment (hayleyavw) [inline kordac/processors/GenericTagBlockProcessor.py]: Indent 43, 45

File: kordac/processors/GenericContainerBlockProcessor.py (hunk 9 above)
Comment (hayleyavw) [inline kordac/processors/GenericContainerBlockProcessor.py]: just read a processor that included:
> Returns:
   Markdown document with comments removed

Should that be here too?...

File: kordac/processors/GenericTagBlockProcessor.py (hunk 11 above)
Comment (hayleyavw) [inline kordac/processors/GenericTagBlockProcessor.py]: Same as earlier:
Should
> Returns:
   Markdown document with comments removed

be here too?...

File: kordac/KordacExtension.py [hunk 12]
```diff
@@ -10,32 +9,48 @@
...
 from jinja2 import Environment, PackageLoader, select_autoescape
 
+
 class KordacExtension(Extension):
+    '''The Kordac markdown extension which enables all the processors,
+    and extracts all the important information to expose externally to
+    the Kordac converter.
+    '''
+
     def __init__(self, processors=[], html_templates={}, extensions=[], *args, **kwargs):
```
Comment (hayleyavw) [inline kordac/KordacExtension.py]: a processor further down defined args and kwargs, should they be defined in this doc string too? or taken out of the other?

File: kordac/processors/ImageBlockProcessor.py [hunk 13]
```diff
@@ -1,49 +1,78 @@
...
         self.required = ext.required_files['images']
-        self.required_parameters = ext.processor_info[self.processor]['required_parameters']
-        self.optional_parameters = ext.processor_info[self.processor]['optional_parameter_dependencies']
 
     def test(self, parent, block):
+        ''' Tests a block to see if the run method should be applied.
+
+        Args:
+            parent: The parent node of the element tree that children
+            will reside in.
```
Comment (hayleyavw) [inline kordac/processors/ImageBlockProcessor.py]: indent 29

File: kordac/processors/ImageBlockProcessor.py [hunk 14]
```diff
@@ -1,49 +1,78 @@
...
 
     def run(self, parent, blocks):
+        ''' Processes the block matching the image pattern, adding
+        any internal images to the KordacExtension result.
+
+        Args:
+            parent: The parent node of the element tree that children
+            will reside in.
+            blocks: A list of strings of the document, where the
+            first block tests true.
```
Comment (hayleyavw) [inline kordac/processors/ImageBlockProcessor.py]: indent 42, 44

File: kordac/processors/InteractiveBlockProcessor.py [hunk 15]
```diff
@@ -53,15 +44,23 @@ def run(self, parent, blocks):
             first block tests true.
```
Comment (hayleyavw) [inline kordac/processors/InteractiveBlockProcessor.py]: returns?

Comment (hayleyavw) [inline kordac/processors/InteractiveBlockProcessor.py]: indent 42, 44

File: kordac/processors/RelativeLinkPattern.py [hunk 16]
```diff
@@ -1,9 +1,9 @@
+from markdown.inlinepatterns import Pattern
 from markdown.util import etree
-import markdown.util as util
-import markdown.inlinepatterns
 import re
 
-class RelativeLinkPattern(markdown.inlinepatterns.Pattern):
+
+class RelativeLinkPattern(Pattern):
```
Comment (hayleyavw) [inline kordac/processors/RelativeLinkPattern.py]: Double quotes used in this docstring - should be single

Comment (hayleyavw) [inline kordac/processors/RelativeLinkPattern.py]: `init` doesn't have a doc string

File: kordac/processors/RelativeLinkPattern.py [hunk 17]
```diff
@@ -24,7 +24,12 @@ def __init__(self, ext, *args, **kwargs):
         self.template = ext.jinja_templates[self.processor]
 
     def handleMatch(self, match):
-
+        '''
```
Comment (hayleyavw) [inline kordac/processors/RelativeLinkPattern.py]: statement about this functions purpose?

File: kordac/processors/RemovePostprocessor.py [hunk 18]
```diff
@@ -1,8 +1,20 @@
...
+    ''' Parses the output document and removes all remove html tags
+    (i.e. <remove> or </remove>) keeping the body of said tags. This
+    allows for the returning of illegal html for further processing.
+    '''
+
     def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
 
     def run(self, text):
+        '''
```
Comment (hayleyavw) [inline kordac/processors/RemovePostprocessor.py]: statement about function's purpose

File: kordac/processors/RemovePostprocessor.py [hunk 19]
```diff
@@ -1,8 +1,20 @@
...
 
+
 class RemovePostprocessor(Postprocessor):
+    ''' Parses the output document and removes all remove html tags
+    (i.e. <remove> or </remove>) keeping the body of said tags. This
+    allows for the returning of illegal html for further processing.
+    '''
+
     def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
```
Comment (hayleyavw) [inline kordac/processors/RemovePostprocessor.py]: missing doc string?

File: kordac/processors/RemoveTitlePreprocessor.py [hunk 20]
```diff
@@ -1,18 +1,38 @@
 from markdown.preprocessors import Preprocessor
 import re
 
+
 class RemoveTitlePreprocessor(Preprocessor):
+    '''Removes the first found title from the given document.'''
```
Comment (hayleyavw) [inline kordac/processors/RemoveTitlePreprocessor.py]: trailing ''' should be on new line

File: kordac/processors/errors/ArgumentMissingError.py [hunk 21]
```diff
@@ -1,17 +1,18 @@
 from kordac.processors.errors.Error import Error
 
-class ParameterMissingError(Error):
+
+class ArgumentMissingError(Error):
     """Exception raised when a custom markdown tag in not matched.
```
Comment (hayleyavw) [inline kordac/processors/errors/ArgumentMissingError.py]: should be single quotes

File: kordac/processors/errors/ArgumentMissingError.py [hunk 22]
```diff
@@ -1,17 +1,18 @@
...
 
     Attributes:
         tag -- tag which was not matched
```
Comment (hayleyavw) [inline kordac/processors/errors/ArgumentMissingError.py]: sudden change in syntax?! should be colon, not double hyphen

File: kordac/processors/errors/InvalidParameterError.py [hunk 23]
```diff
@@ -1,5 +1,6 @@
 from kordac.processors.errors.Error import Error
 
+
 class InvalidParameterError(Error):
     """Exception raised when an invalid parameter value is found.
```
Comment (hayleyavw) [inline kordac/processors/errors/InvalidParameterError.py]: single quotes pls

File: kordac/processors/errors/NoSourceLinkError.py [hunk 24]
```diff
@@ -1,5 +1,6 @@
 from kordac.processors.errors.Error import Error
 
+
 class NoSourceLinkError(Error):
     """Exception raised when no source link is found for a video
```
Comment (hayleyavw) [inline kordac/processors/errors/NoSourceLinkError.py]: single quotes

File: kordac/processors/errors/NoVideoIdentifierError.py [hunk 25]
```diff
@@ -1,5 +1,6 @@
 from kordac.processors.errors.Error import Error
 
+
 class NoVideoIdentifierError(Error):
     """Exception raised when no identifier is found for a video
```
Comment (hayleyavw) [inline kordac/processors/errors/NoVideoIdentifierError.py]: single quote all the things!

File: kordac/processors/utils.py [hunk 26]
```diff
@@ -1,49 +1,145 @@
...
+                context[parameter] = transform(context[parameter])
+    return context
+
+
+def find_transformation(ext, option):
+    '''
+    Returns a transformation for a given string.
+    TODO:
+    In future should be able to combine piped transformations
+    into a single function.
```
Comment (hayleyavw) [inline kordac/processors/utils.py]: what is this magic you speak of?

File: kordac/KordacExtension.py (hunk 12 above)
Comment (ravenmaster001) [inline kordac/KordacExtension.py]: I vote taken out of the other.

File: kordac/processors/GenericContainerBlockProcessor.py (hunk 9 above)
Comment (ravenmaster001) [inline kordac/processors/GenericContainerBlockProcessor.py]: Comment is a preprocessor, which is a different type of processor that takes in lines and returns those lines back with any modifications applied.

File: kordac/processors/GenericTagBlockProcessor.py (hunk 11 above)
Comment (ravenmaster001) [inline kordac/processors/GenericTagBlockProcessor.py]: See reply above.

File: kordac/processors/InteractiveBlockProcessor.py (hunk 15 above)
Comment (ravenmaster001) [inline kordac/processors/InteractiveBlockProcessor.py]: Nope

File: kordac/processors/utils.py (hunk 26 above)
Comment (ravenmaster001) [inline kordac/processors/utils.py]: The output of one function is the input of another. I imagine you'd write them with a pipe ('|') between them.

**Pull request merged**